File: src/frontend/src/services/uazapi.ts
/**
 * Uazapi v2 — frontend service (instance-scoped operations only).
 *
 * All calls that touch the Uazapi API directly require an instance `token`
 * obtained from the backend after POST /api/v1/whatsapp/provision
 * (admintoken — server-side only, never exposed here).
 *
 * Response envelope: { instance: { ... } } — see instancia_uazapi.md §3 and §4.
 */

const UAZAPI_BASE = (import.meta.env.VITE_UAZAPI_URL as string | undefined) ?? '';
const API_BASE    = (import.meta.env.VITE_API_URL    as string | undefined) ?? '/api/v1';

function instanceHeaders(token: string): HeadersInit {
  return { 'Content-Type': 'application/json', token };
}

// ── Shared types ─────────────────────────────────────────────────────────────

export type InstanceStatus =
  | 'connecting'   // waiting for QR scan
  | 'connected'    // paired successfully
  | 'disconnected' // logged out or banned
  | 'qrExpired';   // QR timed out — must call connectInstance again

// ── Backend provision response ────────────────────────────────────────────────

export interface ProvisionResponse {
  /** Uazapi instance token — use for all subsequent calls */
  token: string;
  /** Uazapi internal instance ID */
  instanceId: string;
  /** Sanitised instance name stored on Uazapi (e.g. "optus-clinica") */
  instanceName: string;
}

/**
 * Step 1 (frontend side) — calls the Pulseo backend which creates the Uazapi
 * instance server-side using the admintoken. Returns the instance token for use
 * in connectInstance() and getInstanceStatus().
 *
 * Requires a valid Supabase JWT in the Authorization header so the backend can
 * identify the tenant automatically.
 *
 * Throws if the backend returns a non-2xx response.
 */
export async function provisionInstance(supabaseJwt: string): Promise<ProvisionResponse> {
  const res = await fetch(`${API_BASE}/whatsapp/provision`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${supabaseJwt}`,
    },
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(
      (body as { detail?: string }).detail ??
      `Provision failed: ${res.status}`
    );
  }
  return res.json() as Promise<ProvisionResponse>;
}

// ── Connect — generates QR Code (Step 2, instancia_uazapi.md §4) ─────────────

/**
 * Shape returned by POST /instance/connect (instancia_uazapi.md §4):
 * {
 *   "instance": {
 *     "status": "connecting",
 *     "qrcode": "data:image/png;base64,..."
 *   }
 * }
 */
export interface ConnectResponse {
  /** Data URI ready for <img src>: "data:image/png;base64,..." */
  qrcode: string;
  status: 'connecting';
}

/**
 * Requests a QR Code for the given instance (instancia_uazapi.md §4).
 * Body is intentionally empty — omitting "phone" triggers QR mode (not SMS).
 * Throws if the instance does not exist or the token is invalid.
 */
export async function connectInstance(token: string): Promise<ConnectResponse> {
  const res = await fetch(`${UAZAPI_BASE}/instance/connect`, {
    method: 'POST',
    headers: instanceHeaders(token),
    body: JSON.stringify({}),
  });
  if (!res.ok) {
    throw new Error(`Uazapi connect failed: ${res.status}`);
  }
  const body = await res.json() as {
    instance: { status: string; qrcode: string };
  };
  // Strict shape: { instance: { status: "connecting", qrcode: "data:image/png;base64,..." } }
  return {
    status: 'connecting',
    qrcode: body.instance.qrcode,
  };
}

// ── Status polling (instancia_uazapi.md §4 — Dica de Ouro — poll every 5 s) ──

/**
 * Shape returned by GET /instance/status (instancia_uazapi.md §4):
 * {
 *   "instance": {
 *     "status": "connected" | "connecting" | "disconnected"
 *   }
 * }
 */
export interface StatusResponse {
  status: InstanceStatus;
  /** Present only when status === "connected" */
  profileName?: string;
  isBusiness?: boolean;
  owner?: string;
}

/**
 * Returns current instance status (instancia_uazapi.md §4 — GET /instance/status).
 * When status reaches "connected", profileName / isBusiness / owner are populated.
 * When the QR Code window (~60 s) elapses without a scan, Uazapi returns a
 * non-2xx response — this function normalises that to status "qrExpired" so the
 * UI can show the "Gerar novo QR Code" button.
 */
export async function getInstanceStatus(token: string): Promise<StatusResponse> {
  let res: Response;
  try {
    res = await fetch(`${UAZAPI_BASE}/instance/status`, {
      headers: instanceHeaders(token),
    });
  } catch {
    // Network error — treat as transient, not expired
    throw new Error('Uazapi unreachable');
  }

  if (!res.ok) {
    // Non-2xx from Uazapi during QR window → QR has expired
    return { status: 'qrExpired' };
  }

  const body = await res.json() as {
    instance: {
      status: string;
      profileName?: string;
      isBusiness?: boolean;
      owner?: string;
    };
  };

  // Strict shape: { instance: { status: "connecting" | "connected" | ... } }
  const raw = body.instance;
  const status: InstanceStatus =
    raw.status === 'connected'    ? 'connected'    :
    raw.status === 'connecting'   ? 'connecting'   :
    raw.status === 'disconnected' ? 'disconnected' :
                                    'qrExpired';

  return {
    status,
    profileName: raw.profileName,
    isBusiness:  raw.isBusiness,
    owner:       raw.owner,
  };
}

// ── Anti-ban: delay settings ─────────────────────────────────────────────────

/**
 * Apply message delay to reduce ban risk.
 * "new"    → number < 30 days old: longer delays (5–12 s)
 * "warmed" → established number:   shorter delays (3–8 s)
 */
export async function applyDelaySettings(
  token: string,
  mode: 'new' | 'warmed'
): Promise<void> {
  const body =
    mode === 'new'
      ? { msg_delay_min: 5, msg_delay_max: 12 }
      : { msg_delay_min: 3, msg_delay_max: 8 };
  const res = await fetch(`${UAZAPI_BASE}/instance/updateDelaySettings`, {
    method: 'POST',
    headers: instanceHeaders(token),
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`Uazapi delay settings failed: ${res.status}`);
}

// ── Anti-ban: privacy defaults ───────────────────────────────────────────────

export async function applyPrivacyDefaults(token: string): Promise<void> {
  const res = await fetch(`${UAZAPI_BASE}/instance/privacy`, {
    method: 'POST',
    headers: instanceHeaders(token),
    body: JSON.stringify({
      groupadd:     'contacts',
      last:         'contacts',
      online:       'contacts',
      profile:      'contacts',
      readreceipts: 'all',
      calladd:      'known',
    }),
  });
  if (!res.ok) throw new Error(`Uazapi privacy failed: ${res.status}`);
}

// ── Presence (simulate business hours) ──────────────────────────────────────

export async function setPresence(
  token: string,
  presence: 'available' | 'unavailable'
): Promise<void> {
  const res = await fetch(`${UAZAPI_BASE}/instance/presence`, {
    method: 'POST',
    headers: instanceHeaders(token),
    body: JSON.stringify({ presence }),
  });
  if (!res.ok) throw new Error(`Uazapi presence failed: ${res.status}`);
}

// ── Send text message ────────────────────────────────────────────────────────

export interface SendMessagePayload {
  /** Destination in E.164 format, e.g. "5511999999999" */
  number: string;
  text: string;
}

export async function sendTextMessage(
  token: string,
  payload: SendMessagePayload
): Promise<void> {
  const res = await fetch(`${UAZAPI_BASE}/message/text`, {
    method: 'POST',
    headers: instanceHeaders(token),
    body: JSON.stringify(payload),
  });
  if (!res.ok) throw new Error(`Uazapi send failed: ${res.status}`);
}

File: src/backend/services/uazapi.py
"""
Uazapi HTTP client — centralised integration layer.

All outbound calls to uazapi.dev pass through this module.
Callers receive typed dataclasses; raw HTTP details stay here.
"""

import os
from dataclasses import dataclass
from typing import Any

import httpx

UAZAPI_BASE      = os.environ.get("UAZAPI_BASE_URL", "https://uazapi.dev")
UAZAPI_ADMINTOKEN = os.environ.get("UAZAPI_ADMINTOKEN", "")
SYSTEM_NAME      = os.environ.get("UAZAPI_SYSTEM_NAME", "Pulseo")
_TIMEOUT         = 20


@dataclass
class InstanceInfo:
    token:         str
    instance_id:   str
    instance_name: str


@dataclass
class ConnectionStatus:
    state:        str           # connected | connecting | disconnected | banned | error
    phone_number: str | None
    profile_name: str | None
    owner_jid:    str | None
    is_business:  bool | None
    qrcode:       str | None    # base64 image when state == connecting


class UazapiError(Exception):
    def __init__(self, msg: str, status_code: int = 0):
        super().__init__(msg)
        self.status_code = status_code


class UazapiClient:
    """
    Thin async wrapper around the Uazapi REST API.

    Usage:
        client = UazapiClient()                    # admintoken from env
        client = UazapiClient(instance_token=tok)  # instance-scoped calls
    """

    def __init__(self, instance_token: str | None = None):
        self._instance_token = instance_token

    # ── Internal helpers ──────────────────────────────────────────────────────

    def _admin_headers(self) -> dict[str, str]:
        return {"admintoken": UAZAPI_ADMINTOKEN, "Content-Type": "application/json"}

    def _inst_headers(self) -> dict[str, str]:
        if not self._instance_token:
            raise UazapiError("instance_token required for this operation")
        return {"token": self._instance_token, "Content-Type": "application/json"}

    @staticmethod
    def _check(resp: httpx.Response, context: str) -> dict[str, Any]:
        if resp.status_code not in (200, 201):
            raise UazapiError(
                f"Uazapi {context} failed [{resp.status_code}]: {resp.text[:300]}",
                status_code=resp.status_code,
            )
        return resp.json()

    # ── Admin operations (require admintoken) ─────────────────────────────────

    async def create_instance(self, name: str) -> InstanceInfo:
        """Create a new Uazapi instance. Returns token + ids."""
        async with httpx.AsyncClient(timeout=_TIMEOUT) as http:
            resp = await http.post(
                f"{UAZAPI_BASE}/instance/create",
                headers=self._admin_headers(),
                json={"name": name, "systemName": SYSTEM_NAME},
            )
        data = self._check(resp, "create_instance")
        inst = data.get("instance", {})
        token = inst.get("token", "")
        if not token:
            raise UazapiError("Uazapi did not return instance token")
        return InstanceInfo(
            token=token,
            instance_id=inst.get("id") or inst.get("instanceId", ""),
            instance_name=inst.get("name") or name,
        )

    # ── Instance operations (require instance_token) ──────────────────────────

    async def connect(self) -> ConnectionStatus:
        """
        Request connection / generate QR Code.
        Returns status with base64 qrcode when state == 'connecting'.
        """
        async with httpx.AsyncClient(timeout=_TIMEOUT) as http:
            resp = await http.post(
                f"{UAZAPI_BASE}/instance/connect",
                headers=self._inst_headers(),
                json={},
            )
        data = self._check(resp, "connect")
        inst = data.get("instance", {})
        return ConnectionStatus(
            state=inst.get("status", "connecting"),
            phone_number=inst.get("phone"),
            profile_name=inst.get("profileName"),
            owner_jid=inst.get("ownerJid"),
            is_business=inst.get("isBusiness"),
            qrcode=inst.get("qrcode"),
        )

    async def get_status(self) -> ConnectionStatus:
        """Poll current connection status for this instance."""
        async with httpx.AsyncClient(timeout=_TIMEOUT) as http:
            resp = await http.get(
                f"{UAZAPI_BASE}/instance/status",
                headers=self._inst_headers(),
            )
        data = self._check(resp, "get_status")
        inst = data.get("instance", {})
        return ConnectionStatus(
            state=inst.get("status", "disconnected"),
            phone_number=inst.get("phone"),
            profile_name=inst.get("profileName"),
            owner_jid=inst.get("ownerJid"),
            is_business=inst.get("isBusiness"),
            qrcode=inst.get("qrcode"),
        )

    async def send_text(self, phone: str, message: str) -> dict[str, Any]:
        """Send a plain-text WhatsApp message."""
        async with httpx.AsyncClient(timeout=_TIMEOUT) as http:
            resp = await http.post(
                f"{UAZAPI_BASE}/message/sendText",
                headers=self._inst_headers(),
                json={"phone": phone, "message": message},
            )
        return self._check(resp, "send_text")

    async def configure_webhook(self, url: str, events: list[str]) -> None:
        """Register the webhook URL and subscribed event types."""
        async with httpx.AsyncClient(timeout=_TIMEOUT) as http:
            resp = await http.post(
                f"{UAZAPI_BASE}/webhook",
                headers=self._inst_headers(),
                json={"url": url, "events": events},
            )
        self._check(resp, "configure_webhook")

    async def configure_delay(self, min_s: int, max_s: int) -> None:
        """Set anti-ban message delay (seconds)."""
        async with httpx.AsyncClient(timeout=_TIMEOUT) as http:
            resp = await http.post(
                f"{UAZAPI_BASE}/instance/updateDelaySettings",
                headers=self._inst_headers(),
                json={"msg_delay_min": min_s, "msg_delay_max": max_s},
            )
        self._check(resp, "configure_delay")

    async def configure_privacy(self, settings: dict[str, str]) -> None:
        """Apply WhatsApp privacy settings to the instance."""
        async with httpx.AsyncClient(timeout=_TIMEOUT) as http:
            resp = await http.post(
                f"{UAZAPI_BASE}/instance/privacy",
                headers=self._inst_headers(),
                json=settings,
            )
        self._check(resp, "configure_privacy")

    async def configure_warmup_delay(self, min_s: int, max_s: int) -> None:
        """
        Apply warm-up-aware message delay to the instance.

        Delegates to configure_delay — exists as a named method so callers
        can express intent (warm-up schedule) instead of raw seconds.
        """
        await self.configure_delay(min_s=min_s, max_s=max_s)
